writer.go

```diff
@@ -27,6 +28,7 @@ type Writer struct {
 
 // WriterConfig is a configuration type used to create new instances of Writer.
 type WriterConfig struct {
+	NewPartitionWriter func(partition int, config WriterConfig) PartitionWriter
```

Comment:
Is the reason we need this only to support mocking during tests? If that's the case I'd advocate for making it an unexported field so we don't expose implementation details to the public API.